### PR TITLE
Link var in type restrictions to typespecs.html#defining-a-specification

### PR DESCRIPTION
**Pull request by KronicDeth**
Related to elixir-ecto/ecto#3756.  It would fix it once ecto is updated to use an `ex_doc` with this change.

# Reason for change

The use of `var` in `@callback checkout(adapter_meta, config :: Keyword.t(), (() -> result)) :: result when result: var` can be used to mean `result` has no restrictions, but that `var` unlike real typespec types didn't link anywhere, so reading the Ecto docs, I couldn't figure out if it was an unresolved type or a built-type that wasn't linking.  @josevalim in elixir-ecto/ecto#3756 pointed to https://hexdocs.pm/elixir/typespecs.html#defining-a-specification, so we agreed on the plan to link `var` to there.

# Potential solutions

I don't terribly like this solution I came up with for the linking because it ends up having to duplicate a lot of`Macro.to_string/1`.  To cover having to duplicate more of `Macro.to_string`, I added more tests.  **I'm open to a smaller, more focused change.**

## Alternative solution I tried
* Use a regex to find the `var` and link it.  Didn't work because I couldn't find a regex that would link `var` at the top-level, but not nested when there was a complex nested type in the `when`.  
* I tried using `Macro.to_string/2` instead of custom walking code, but the problem is the callback is passed the AST bottom up, which means it would get the AST for `t()` when the type that needed to be linked is `AutolinkTest.Foo.t()`.

# Testing
* `ex_doc` unit tests
* Used this change as a path dependency for local ecto and got fix in generated docs.
  
  Screen Shot 2021-10-11 at 4 10 20 PM
  Screen Shot 2021-10-11 at 4 10 40 PM

# Changelog
## Enhancements
* Link var in type restrictions to typespecs.html#defining-a-specification
  * On hover show "No restriction"

### CODE REVIEW

**Review by wojtekmach — COMMENTED** [review]
> I don't terribly like this solution I came up with for the linking because it ends up having to duplicate a lot of Macro.to_string/1

Yeah, that is my biggest worry that we're basically reimplementing Macro.to_string/1 and that is a lot of work and a lot of edge cases. At the same time, I think this solution, even if more work, is actually a much nicer one that the existing one (the big regex and counting parens for auto linking).

We have yet another solution, https://github.com/elixir-lang/ex_doc/blob/94cf4040c54877cfa0a67a05a6943ec24c1055a2/lib/ex_doc/language/erlang.ex#L418:L445, and our hope was the basic approach is generic enough that it could be applied to both Erlang & Elixir (and others) type specs, and that we could re-use as much code as possible. I still think that might be the case, just didn't get around trying that. If you have any feedback about that particular solution, it'd be appreciated.

Taking a big step back, and apologies if it is stating the obvious and redundant, but basically we want to take AST and pretty-print it, replacing particular strings with links, while preserving the text layout, the spaces and new lines. Sort of what Macro.to_string/2 did (now deprecated) but much more nicely printed, i.e. using the formatter.

So yeah, not sure what is the best way forward, but I hope the extra context is useful.

File: lib/ex_doc/language/elixir.ex [hunk 1]
```diff
@@ -182,17 +182,226 @@ defmodule ExDoc.Language.Elixir do
...
+    autolink_spec_head(head, config)
+  end
+
+  defp autolink_spec_head({unary_operator, _, [operand]}, config)
+       when unary_operator in [:+, :-] do
+    Atom.to_string(unary_operator) <> autolink_spec_body(operand, config)
+  end
+
+  defp autolink_spec_head({binary_operator, _, [left, right]}, config)
+       when binary_operator in [:+, :-, :<, :<=, :>=, :>] do
```
[the remark this answers is not in the document]
Comment (KronicDeth) [reply lib/ex_doc/language/elixir.ex:211]: Yeah or else the specs for `Kernel` won't look right.

**Comment by KronicDeth** [comment]
Using `Macro.prewalk` is probably the better approach since walking the AST is baked in unlike my solution of custom functions. Still have the issue of the actual string representation being in both with pretty-print helper    

https://github.com/elixir-lang/ex_doc/blob/94cf4040c54877cfa0a67a05a6943ec24c1055a2/lib/ex_doc/language/erlang.ex#L573-L579

I'll attempt a solution closer to that and see if it better in a separate PR and then we can compare.

**Comment by josevalim** [comment]
We will close this for now due to the complexity and the limited benefit. Thank you!

**Pull request closed without merging**